content/blog/custom-field-plugins.md

```diff
@@ -0,0 +1,144 @@
+
+## 1. Add the Field Plugin to the CMS
+
+To register the custom field as a plugin with the CMS, we’ll need to head to the file where we can access the cms instance. In our Next.js demo, we’ll look at `_app.js`. This it the place where we override the default Next app to wrap `Tina` around each page.
+
+```js
+import React from 'react'
+import App from 'next/app'
+import { Tina, TinaCMS } from 'tinacms'
+import { GitClient } from '@tinacms/git-client'
+/*
+ ** 1. Import the custom field component
+ */
+import RangeInput from '../components/RangeInput'
+
+export default class Site extends App {
+  constructor() {
+    super()
+    this.cms = new TinaCMS({
+      apis: {
+        git: new GitClient('http://localhost:3000/___tina'),
+      },
+      sidebar: {
+        position: 'overlay',
+        hidden: process.env.NODE_ENV === 'production',
+      },
+    })
+
+    /*
+     ** 2. Define the field plugin
+     */
+    this.customFieldPlugin = {
+      name: 'range-input',
+      Component: RangeInput,
+    }
+    /*
+     ** 3. Register the plugin with the cms
+     */
+    this.cms.fields.add(this.customFieldPlugin)
+  }
+
+  render() {
+    const { Component, pageProps } = this.props
+    return (
+      <Tina cms={this.cms}>
+        <Component {...pageProps} />
+      </Tina>
+    )
+  }
+}
+```
+
+You’ll want to import the custom field component and then register the plugin with the CMS directly. Notice how we import the `RangeInput` component created in the [previous post](https://tinacms.org/blog/custom-field-components), this is the custom _field component_ that we're now attaching to a _field plugin_.
+
+> If you’re working with Gatsby, this [looks slightly different](https://tinacms.org/docs/gatsby/custom-fields/). _Hint_: you’ll head to the **gatsby-browser.js** file to access the CMS instance.
+
+### Field Plugin Definition
+
+A plugin is used to extend functionality in the CMS. A field plugin in particular allows us to create and register custom fields to the CMS.
+
+The field plugin definition **requires (at a minimum) a name and a component.** The name will be used to reference the custom field in form definitions. The component is what will actually be rendered in the sidebar.
+
+```ts
+interface FieldPlugin {
+  __type: 'field'
+  name: string
+  Component: React.FC<any>
+  type?: string
+  validate?(
+    value: any,
+    allValues: any,
+    meta: any,
+    field: Field
+  ): string | object | undefined
+  parse?: (value: any, name: string, field: Field) => any
+  format?: (value: any, name: string, field: Field) => any
+  defaultValue?: any
+}
+```
+
+You can see that there are additional configuration functions and options. These options are incredibly useful for creating more complex fields that require [validation](https://tinacms.org/docs/fields/custom-fields#validate-optional) or formatting, such as an [email field](https://tinacms.org/docs/gatsby/custom-fields/).
```

Comment:
... finally here is the explanation 😄We should sum up this in the intro to give a sense of purpose to the article.

Reply:
Note that it's still possible to put validate, parse, and format on other fields. The usefulness of FieldPlugins comes from not having to do that every single time. 

For instance, why do we add a "date" field plugin instead of just making the `DateField` component available? Because we want to make sure that the `parse` and `format` functions are always supplied.

```ts
{ 
  name: "created_at", 
  component: "date" 
}
```
is much nicer than:

```ts
import { DateField, parseDate, formatDate} from "tinacms"

{ 
  name: "created_at", 
  component: DateField, 
  parse: parseDate,
  format: formatDate,
},
```

An additional note (that might be too heady for this blog post) is that this approach better adheres to the Dependency Inversion Principle. By using the plugin API to reference the "date" field, the modules that describe the forms do not rely on some concrete implementation of a date field. If you find a new date field, you just update or replace the "date" plugin and it updates all your forms.